client/directclient.go

```diff
@@ -83,67 +85,205 @@ func NewProxyDirectClient(policyList conf.PolicyList) (*ProxyDirectClient, error
+	failures     []int
+	rc           []int
+	requestCount int
+	replicaCount int
+	responsec    chan *respRec
+	cancel       chan struct{}
+}
+
+func (q *stdQuorumer) launchRequest() bool {
+	// step through primaries, then go to handoffs.
+	launch := func(index int, dev *ring.Device) {
+		resp := q.makeRequest(index, dev)
+		select {
+		case q.responsec <- &respRec{i: index, r: resp, d: dev}:
+		case <-q.cancel:
+		}
 	}
-	quorum := int(math.Ceil(float64(len(reqs)) / 2.0))
-	responseClasses := []int{0, 0, 0, 0, 0, 0}
-	responseCount := 0
-	var chosenResponse *http.Response
-	for response := range responses {
-		responseCount++
-		class := response.StatusCode / 100
-		if class <= 5 {
-			responseClasses[class]++
-			if responseClasses[class] >= quorum {
-				chosenResponse = response
-				break
+	if q.requestCount < len(q.devs) {
+		go launch(q.requestCount, q.devs[q.requestCount])
+	} else {
+		var index int // assign handoffs the index of a failed primary, so we send update headers.
+		index, q.failures = q.failures[len(q.failures)-1], q.failures[:len(q.failures)-1]
+		dev := q.more.Next()
+		if dev == nil {
+			return false
+		}
+		go launch(index, dev)
+	}
+	q.requestCount++
+	return true
+}
+
+func (q *stdQuorumer) start() {
+	for q.requestCount < q.replicaCount {
+		q.launchRequest()
+	}
+}
+
+func (q *stdQuorumer) addResponse(index int, resp *http.Response) {
+	q.responses = append(q.responses, resp)
+	if resp.StatusCode >= 500 {
+		q.failures = append(q.failures, index)
+	} else if resp.StatusCode >= 0 {
+		q.rc[resp.StatusCode/100]++
+	}
+}
+
+func (q *stdQuorumer) quorumPossible() bool {
+	// if quorum has become impossible, launch requests until it's possible again.
+	for q.requestCount < q.replicaCount*2 {
+		for i := 2; i < 5; i++ {
+			if (q.rc[i]+q.requestCount)-len(q.responses) >= q.q {
+				return true
 			}
 		}
+		if !q.launchRequest() {
+			break
+		}
+	}
+	return false
+}
+
+func (q *stdQuorumer) quorumMet(writers int) bool {
+	if writers >= q.replicaCount {
+		return true
+	}
+	if writers >= q.q && (len(q.responses)+writers) >= q.replicaCount*2 {
+		return true
+	}
+	for i := 2; i < 5; i++ {
+		if q.rc[i] >= q.q {
+			return true
+		}
 	}
-	// Give any pending requests *some* chance to finish. This will increase
-	// the likelihood that a read immediately after a write will get the latest
-	// information, as we'll force the caller to wait until all backends have
-	// responded, with a timeout limit.
-	timeout := time.After(PostQuorumTimeoutMs * time.Millisecond)
-waiting:
-	for responseCount < len(reqs) {
+	return false
+}
+
+func (q *stdQuorumer) getResponse(timeout time.Duration) *http.Response {
+	end := time.After(timeout)
+	for {
+		// see if quorum has already been met
+		for _, r := range q.responses {
+			if q.rc[r.StatusCode/100] >= q.q {
+				// Give pending requests a chance to finish, to improve consistency of read-after-write.
+				end := time.After(PostQuorumTimeoutMs * time.Millisecond)
```

Comment:
I dislike shadowed vars, but just a suggestion to rename this "end" to something else.